tests/functional/api/groups/members_test.py

```diff
@@ -62,179 +62,135 @@ def test_it_returns_empty_list_if_no_members_in_group(self, app):
 
         assert res.json == []
 
-    @staticmethod
-    def authorization_header(token) -> dict:
-        """Return an Authorization header for the given developer token."""
-        return {"Authorization": "Bearer {}".format(token.value)}
-
 
 class TestAddMember:
-    def test_it_returns_http_204_when_successful(
-        self, app, auth_client, db_session, factories
+    def test_it(self, db_session, group, user, do_request):
+        do_request()
+
+        assert user in group.members
+
+    def test_it_does_nothing_if_the_user_is_already_a_member_of_the_group(
+        self, group, user, do_request
     ):
-        third_party_user = factories.User(authority="thirdparty.com")
-        third_party_group = factories.Group(authority="thirdparty.com")
-        db_session.commit()
+        group.memberships.append(GroupMembership(user=user))
 
-        res = app.post_json(
-            "/api/groups/{pubid}/members/{userid}".format(
-                pubid=third_party_group.pubid, userid=third_party_user.userid
-            ),
-            headers=self.authorization_header(auth_client),
-        )
+        do_request()
 
-        assert res.status_code == 204
+        assert user in group.members
 
-    def test_it_adds_member_to_group(self, app, auth_client, db_session, factories):
-        third_party_user = factories.User(authority="thirdparty.com")
-        third_party_group = factories.Group(authority="thirdparty.com")
-        db_session.commit()
+    def test_it_errors_if_the_pubid_is_unknown(self, do_request):
+        do_request(pubid="UNKNOWN_PUBID", status=404)
 
-        app.post_json(
-            "/api/groups/{pubid}/members/{userid}".format(
-                pubid=third_party_group.pubid, userid=third_party_user.userid
-            ),
-            headers=self.authorization_header(auth_client),
-        )
+    def test_it_errors_if_the_userid_is_unknown(self, do_request, authclient):
+        do_request(userid="acct:UNKOWN_USERNAME@{authclient.authority}", status=404)
 
-        assert third_party_user in third_party_group.members
+    def test_it_errors_if_the_userid_is_invalid(self, do_request):
+        do_request(userid="INVALID_USERID", status=404)
 
-    def test_it_ignores_forwarded_user_header(
-        self, app, factories, db_session, auth_client
-    ):
-        third_party_user = factories.User(authority="thirdparty.com")
-        third_party_group = factories.Group(authority="thirdparty.com")
-        headers = self.authorization_header(auth_client)
-        user2 = factories.User(authority="thirdparty.com")
-        db_session.commit()
```

Comment:
This test was completely wrong. It uses `third_party_user.userid` in both the URL and the `X-Forwarded-User` header and does not include `user2` in the request at all. And then asserts that `user2` was not added to the group. It's been replaced with a correct test.